tools/api-linter/src/visitor.rs

```diff
@@ -0,0 +1,508 @@
+            ItemEnum::AssocConst { .. } => unimplemented!("visit_item ItemEnum::AssocConst"),
+            ItemEnum::AssocType { bounds, default } => {
+                path.push(ComponentType::AssocType, item);
+                if let Some(typ) = default {
+                    self.visit_type(&path, &ErrorLocation::AssocType, typ)?;
+                }
+                self.visit_generic_bounds(&path, bounds)?;
+            }
+            ItemEnum::Constant(constant) => {
+                path.push(ComponentType::Constant, item);
+                self.visit_type(&path, &ErrorLocation::Constant, &constant.type_)?;
+            }
+            ItemEnum::Enum(enm) => {
+                path.push(ComponentType::Enum, item);
+                self.visit_generics(&path, &enm.generics)?;
+                for id in &enm.impls {
+                    self.visit_impl(&path, self.item(id)?)?;
+                }
+                for id in &enm.variants {
+                    self.visit_item(&path, self.item(id)?)?;
+                }
+            }
+            ItemEnum::ForeignType => unimplemented!("visit_item ItemEnum::ForeignType"),
+            ItemEnum::Function(function) => {
+                path.push(ComponentType::Function, item);
+                self.visit_fn_decl(&path, &function.decl)?;
+                self.visit_generics(&path, &function.generics)?;
+            }
+            ItemEnum::Import(import) => {
+                if let Some(id) = &import.id {
+                    path.push_raw(ComponentType::ReExport, &import.name, item.span.as_ref());
+                    self.check_external(&path, &ErrorLocation::ReExport, id)
+                        .context(here!())?;
+                }
+            }
+            ItemEnum::Method(method) => {
+                path.push(ComponentType::Method, item);
+                self.visit_fn_decl(&path, &method.decl)?;
+                self.visit_generics(&path, &method.generics)?;
+            }
+            ItemEnum::Module(module) => {
+                if !module.is_crate {
+                    path.push(ComponentType::Module, item);
+                }
+                for id in &module.items {
+                    let module_item = self.item(id)?;
+                    // Re-exports show up twice in the doc json: once as an `ItemEnum::Import`,
+                    // and once as the type as if it were originating from the root crate (but
+                    // with a different crate ID). We only want to examine the `ItemEnum::Import`
+                    // for re-exports since it includes the correct span where the re-export occurs,
+                    // and we don't want to examine the innards of the re-export.
+                    if module_item.crate_id == self.root_crate_id {
+                        self.visit_item(&path, module_item)?;
+                    }
+                }
+            }
+            ItemEnum::OpaqueTy(_) => unimplemented!("visit_item ItemEnum::OpaqueTy"),
```

Comment:
I wasn't able to come up with a piece of code that would trigger this, so I opted to leave it unimplemented until we have an example of it. It seems unlikely to come up since none of our runtime crates or generated code triggered it though.